File: RobotsJava/TheBase.java

/*
//------------------------------------------------------------//

**************		****		****		**************		
**************		****		****		**************
****		 ****		****		****		****
****		 ****		****		****		****
**************		****		****		****
**************		****		****		****
****					****		****		****
****					****		****		****
****					*************		**************
****					*************		**************		

-------------------------------------------------------------//
Programmer				Date				Modification Reason

Jaime Alvarez			12-21-13			Initial Implementation


//-----------------------------------------------------------//
Description:
This class represents the base of all robots. This class will drive the base motors,
receive encoder data, and send data to the server. 




*/
import java.io.*;
import java.net.*;

class TheBase
{
	//The X, Y, and Phi coordinates (X,Y,PHI) initially set to zero.
	double Xcurrent = 0;
	double Ycurrent = 0;
	double PHIcurrent = 0;
	
	//Length of wheel base (inches)
	double lengthOfWheelBase = 15;
	
	//These variabels keep track of the number of pulses received during a time interval
	int LWPulses = 0;
	int RWPulses = 0;
	
	//The IP address of the server
	String serverAddress = "172.0000000";
	
	//Constructors
	//This constructor should be called for creating the robot that will be centered at the origin i.e. Quarterback
	public TheBase()
	{
	}
	
	//This constructor should be called for creating a robot in reference to the origin i.e. Receiver
	public TheBase(double x, double y, double phi)
	{
		Xcurrent = x;
		Ycurrent = y;
		PHIcurrent = phi;
	}
	
	//Method to setup network connection
	public boolean connectWithServer()
	{
		//Set up socket with server ip address and port number.
		//Also I/O streams
		try(
				Socket s = new Socket(serverAddress, 4465);
				PrintWriter out = new PrintWriter(s.getOutputStream(), true);
				BufferedReader in = new BufferedReader(new InputStreamReader(s.getInputStream()));
				)
				{
					//logic
					
					//connection successful so return true
					return true;
				} 		
		catch(IOException e)
		{
			System.err.println("Couldn't get I/O for the connection to the server " + e.getMessage());
			return false;
			
		}
		catch(Exception e1)
		{
			System.err.println(e1.getMessage());
			return false;
		}
		
	}	
	
	
	//This method will take the data received from the server and determine in what direction the robot
	//will move.  It will also use the given speeds for the left and right motors determined by the server.
	//For example: User pushes the thumbstick left up, the server will determine how much should the robot will turn
	//and move forward appropriately.
	public void driveBaseMotors(String direction, int leftMotorSpeed, int rightMotorSpeed)
	{
		if(direction == "Full Forward")
		{
			//Drive both motors in forward direction (same speed)
		}
		else if(direction == "Full Backward")
		{
			//Drive both motors in backward direction (same speed)
		}
		else if(direction == "Rotate Left")
		{
			//Have right motor drive forward and the left motor drive backward
		}
		else if(direction == "Rotate Right")
		{
			//Have right motor drive backward and the left motor drive forward 
		}
		else if(direction == "Forward Left")
		{
			//Have right motor drive faster forward than the left motor (also going forward)
		}
		else if(direction == "Forward Right")
		{
			//Have left motor drive faster forward than the right motor (also going forward)
		}
		else if(direction == "Backward Left")
		{
			//Have right motor drive faster backward than the left motor (also going backward)
		}
		else if(direction == "Backward Right")
		{
			//Have left motor drive faster backward than the right motor (also going backward)
		}
		else
		{
			//Do nothing
		}		
	}
	
	//These next two method checks how many pulses are received at a certain time interval from the encoders
	//and will save the number of pulses to LWPulses and RWPulses variables.
	public void getLeftWheelPulses()
	{
		try
		{
			//Logic to get pulses from left encoder
			//LWPulses = ......
		}
		catch(Exception ex)
		{
			//Handle any errors
		}
		
	}
	
	public void getRightWheelPulses()
	{
		try
		{
			//Logic to get pulses from right encoder
			//RWPulses = ......
		}
		catch(Exception ex)
		{
			//Handle any errors
		}
		
	}
	
	//This method will use the number of pulses received during a time interval (using LWPulses and RWPulses variables)
	//to update the position of the robot.
	public void updatePosition()
	{
		//Delta Distance Right Wheel (dDRW), Delta Distance Left wheel (dDLW), and Delta Distance Center of Robot (dDCR) - (initially zero change)
		double dDRW = 0;
		double dDLW = 0;
		double dDCR = 0;
		
		//Get number of pulses for both left and right encoders and determine distance traveled		
		dDLW = 0.012 * LWPulses;
		dDRW = 0.012 * RWPulses;
		
		//Determine the distance traveled with respect to the center of the robot
		dDCR = (dDLW + dDRW)/2;
		
		//Update orientation of robot first 
		PHIcurrent = PHIcurrent + (dDRW - dDLW) / lengthOfWheelBase;
		
		//Then X and Y coordinates
		Xcurrent = Xcurrent + dDCR*Math.sin(-PHIcurrent);
		Ycurrent = Ycurrent + dDCR*Math.cos(PHIcurrent);
		
		//Clean up delta changes
		dDRW = dDLW = dDCR = 0;
		
		//Send new coordinates to server
		sendCoordinates();		
		
	}
	
	//This method will send the current (X,Y,PHI) coordinates to the server
	public void sendCoordinates()
	{
		//Logic to send data
	}

}

File: BeagleBonePythonCode/TheBase.py
'''//------------------------------------------------------------//

**************		****	 ****		**************		
**************		****	 ****		**************
****	  ****		****	 ****		****
****	  ****		****	 ****		****
**************		****	 ****		****
**************		****	 ****		****
****			****	 ****		****
****			****     ****		****
****			*************		**************
****			*************		**************		

-------------------------------------------------------------//
Programmer	    Date		Modification Reason

Jaime Alvarez	    01-31-14		Initial Implementation


//-----------------------------------------------------------//

Description:
This class will simulate the base of each robot.
It provides functions as driving the motors,
collect encoder position data, etc.
'''
import socket
import math
import Adafruit_BBIO.UART as UART
import serial

class TheBase:
    #The X, Y, and Phi coordinates (X,Y,PHI) initially set to zero.
    Xcurrent = 0.0
    Ycurrent = 0.0
    PHIcurrent = 0.0

    #Length of wheel base (inches)
    lengthOfWheelBase = 15

    #These variabels keep track of the number of pulses received during a time interval
    LWPulses = 0
    RWPulses = 0

    #Set up the base motors
    #Ser1 is assuming to be the left motor
    #Ser2 is assuming to be the right motor
    UART.setup("UART1")
    UART.setup("UART2")
    ser1 = serial.Serial(port = "/dev/ttyO1", baudrate=9600)
    ser2 = serial.Serial(port = "/dev/ttyO2", baudrate=9600)
    ser1.close()
    ser2.close()
    ser1.open()
    ser2.open()

    #Server IP, port, and socket to send position data to
    serverIP = "127.0.0.1"
    serverPort = 4489
    Disconnected = False
    serverSocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    #Socket to listen for commands from the server
    theSocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    

    #Constructor
    #If for constructing the Quarterback, create the object without any parameters since it'll be the reference
    #If for any other robot, such as the Receiver, then type in the coordinates it'll start off with respect to the Quarterback
    def __init__(self, x = 0.0, y = 0.0, phi = 0.0):
        self.Xcurrent = x
        self.Ycurrent = y
        self.PHIcurrent = phi

    #Displays the robot's current coordinates
    def displayCoordinates(self):
        print self.Xcurrent
        print self.Ycurrent
        print self.PHIcurrent

    #This method will take a the speed data received from the socket
    #And from it, it will differentiate if it will go straight forward
    #Or backwards. It'll also know when to turn and it which direction
    #For going straight forward or backward, the data will contain a value
    #in the format of 'S255'. While turning, the format will be 'TS255'
    def driveBaseMotors(self, speedData):
        if self.ser1.isOpen() and self.ser2.isOpen():
	    if speedData[0] == "S":
                #Go straight backward or forward
                #Logic to drive motors, one is reversed which is right one
                i = 0
                while True:
                    if speedData[i] == "X":
                        break
                    else:
                        i += 1
                self.ser1.write(speedData[0:i] + "\n\r")
                self.ser2.write("S" + str(-1*int(speedData[i+1:])) + "\n\r" )
            elif speedData[0] == "T":
                #Turn robot according to how much the thumbstick is pushed
                #Logic to turn
                #Logic for turning clockwise
                if int(speedData[1:]) > 0:
                    self.ser1.write("S" + speedData[1:] + "\n\r")
                    self.ser2.write("S0 \n\r")
                else:
                    #Logic for turning counterclockwise
                    self.ser1.write("S0 \n\r")
                    self.ser2.write("S" + speedData[1:] + "\n\r")
        elif self.ser1.isOpen() and not self.ser2.isOpen():
            print "Ser1 is opened but not Ser2!"
        elif not self.ser1.isOpen() and self.ser2.isOpen():
            print "Ser2 is opened but not Ser1!"
        else:
            print "Both Ser1 and Ser2 are not opened!"

    #These next two method checks how many pulses are received at a certain time interval from the encoders
    #and will save the number of pulses to LWPulses and RWPulses variables
    def getLeftWheelPulses(self):
        #Try to get pulses from the Left encoder
        try:
            #Logic to do it, LWPulses = .....
            self.LWPulses = 5
            print "Got Left Wheel Pulse"
        except Exception as inst:
            print type(inst)
            print inst.args
            print inst
        else:
            #Anything else to do with encoder
            doen="Done"

    #For right wheel
    def getRightWheelPulses(self):
        #Try to get pulses from the Right encoder
        try:
            #Logic to do it, RWPulses = .....
            self.RWPulses = 5
            print "Got Right Wheel Pulse"
        except Exception as inst:
            print type(inst)
            print inst.args
            print inst
        else:
            #Anything else to do with encoder
            doen="Done"

    #This method will use the number of pulses received during a time interval (using LWPulses and RWPulses variables)
    #to update the position of the robot
    def updatePosition(self):
        #Delta Distance Right Wheel (dDRW), Delta Distance Left wheel (dDLW), and Delta Distance Center of Robot (dDCR) - (initially zero change)
        dDRW = 0.0
        dDLW = 0.0
        dDCR = 0.0
		
        #Get number of pulses for both left and right encoders and determine distance traveled		
        dDLW = 0.012 * self.LWPulses
        dDRW = 0.012 * self.RWPulses
	
        #Determine the distance traveled with respect to the center of the robot
        dDCR = (dDLW + dDRW)/2
		
        #Update orientation of robot first 
        self.PHIcurrent = self.PHIcurrent + (dDRW - dDLW) / self.lengthOfWheelBase
		
        #Then X and Y coordinates
        self.Xcurrent = self.Xcurrent + dDCR*math.sin(-self.PHIcurrent)
        self.Ycurrent = self.Ycurrent + dDCR*math.cos(self.PHIcurrent)
		
        #Clean up delta changes
        dDRW = dDLW = dDCR = 0

        #Testing
        self.Xcurrent += .0003
        self.Ycurrent += .0001
        self.PHIcurrent += .0002
		
        #Send new coordinates to server
        self.sendCoordinates()


    #This method will send the current (X,Y,PHI) coordinates to the server
    def sendCoordinates(self):
        #Logic to send coordinates
        coordinates = "{0:.4f}".format(self.Xcurrent) + "X" + "{0:.4f}".format(self.Ycurrent) + "X" + "{0:.4f}".format(self.PHIcurrent) + "X"
        self.serverSocket.sendto(coordinates, (self.serverIP, self.serverPort))
        
    #This method will set up the IP and port the robot will be listening from
    def setupListenSocket(self, ip, port):
        self.theSocket.bind((ip, port))

    def setupSendSocket(self, ip, port):
        self.serverIP = ip
        self.serverPort = port

    #This method will retreive the command from the server
    def retreiveCommands(self):
        data, addr = self.theSocket.recvfrom(32)
        if data[0] == "S" or data[0] == "T":
            self.driveBaseMotors(data)
        elif data[0] == "B":
            print "Button: ", data[1:]
        elif data[0] == "D":
            self.Disconnected = True
        else:
            print "Unknown data"

    
        
        

    
    
